Implement Snapcast grouping with standard HA actions

mill1000 opened the pull request:
With the release of 2025.6 and it's [improved support of joining media players](https://www.home-assistant.io/blog/2025/06/11/release-20256/#joinunjoin-groups-of-media-players), I thought it would be timely to finally make this PR.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
This PR deprecates the custom `snapcast.join` and `snapcast.unjoin` actions and adds the standard HA actions `media_player.join` and `media_player.unjoin`. When deprecated actions are used a repair issue is created to alert users.

Additionally, to align with the HA architecture, Snapcast group entities are deprecated and repair issues are created to alert users when any action is called on a media player entity representing a group.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request:  https://github.com/home-assistant/home-assistant.io/pull/39568
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository] 

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

mill1000 committed:
Implement Snapcast grouping in accordance with latest HA architecture

- Define new config option to enable group entities
- Add handlers for standard media player join/unjoin services
- Add GROUPING to supported features for client device
- Remove custom join/unjoin services and related strings, icons, etc

joostlek reviewed (changes requested):
Let's take one step back

1. We should have the existing services deprecated for 6 months by raising a repair issue once used explaining the user what they should change.
2. Why don't we create group entities now? They can still have a use case right?

mill1000 committed:
Restore legacy Snapcast grouping actions

mill1000 committed:
Create repair issues when legacy grouping actions are used

mill1000 commented:
> * We should have the existing services deprecated for 6 months by raising a repair issue once used explaining the user what they should change.

Got it. I've restored the original actions and modified them to create a repair issue if used.

> * Why don't we create group entities now? They can still have a use case right?

Ah, based on my reading of https://github.com/home-assistant/architecture/issues/364 specifically the proposal here https://github.com/home-assistant/architecture/issues/364#issuecomment-625405413 I understood creating entities for dynamic groups to be undesirable.

joostlek commented:
> Ah, based on my reading of https://github.com/home-assistant/architecture/issues/364 specifically the proposal here https://github.com/home-assistant/architecture/issues/364#issuecomment-625405413 I understood creating entities for dynamic groups to be undesirable.

I did not know about this discussion. But I think if we decide to remove them, we should completely remove them instead of putting them behind a feature flag. Let me ask in our core meeting if we still share the same opinion as back then

luar123 commented:
I agree, for most users the group entities are probably useless, because they are dynamic and each time you change the grouping old groups could get deleted or new ones added. So not adding group entities makes sense as a default behavior.
However, some users might use snapcast with static groups, once defined and never touched. Those users might use the group entities instead of client entities. So in my view it makes sense to allow adding them optionally.

mill1000 committed:
Update unit tests to pass with grouping and new config options

joostlek commented:
So people using static groups, can they still use the normal clients? As in, we could in theory deprecate the groups and raise issues when people use them in automations or scripts or use their actions, and in 6 months remove them.

mill1000 committed:
Add unit tests for grouping

luar123 commented:
Technically, in snapcast the stream/source is a group property, however, we propagate it to the clients. So when a client switches the source the whole group switches. 
And for the volume you can always change the volume of the source, mpd, Spotify, etc.
So i guess removing group entities won't be a big loss.

mill1000 committed:
Deprecate Snapcast group entities

Create issues when services are called on group entities.
Remove integration option for enabling group entites.

mill1000 committed:
Remove leftover await

joostlek commented on homeassistant/components/snapcast/__init__.py:
```diff
@@ -24,6 +24,9 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
+    # Reload entry when its updated
+    entry.async_on_unload(entry.add_update_listener(async_reload_entry))
+
```
not needed anymore

mill1000 replied:
Got it.

joostlek commented on homeassistant/components/snapcast/__init__.py:
```diff
@@ -34,3 +37,8 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
         # disconnect from server
         await snapcast_data.disconnect()
     return unload_ok
+
+
+async def async_reload_entry(hass: HomeAssistant, entry: ConfigEntry) -> None:
+    """Reload a config entry."""
+    await hass.config_entries.async_reload(entry.entry_id)
```
idem

joostlek commented on homeassistant/components/snapcast/config_flow.py:
```diff
@@ -47,6 +47,7 @@ async def async_step_user(self, user_input=None) -> ConfigFlowResult:
             else:
                 client.stop()
                 return self.async_create_entry(title=DEFAULT_TITLE, data=user_input)
+
```
```suggestion
```
For the sake of touched files

joostlek commented on homeassistant/components/snapcast/strings.json:
```diff
@@ -58,5 +58,15 @@
+    },
+    "deprecated_group_entities": {
+      "title": "Snapcast Groups Entities Deprecated",
+      "description": "Snapcast group entities are deprecated and will be removed in 2026.2.12."
```
```suggestion
      "description": "Snapcast group entities are deprecated and will be removed in 2026.2. Please use the 'media_player.join' and 'media_player.unjoin' actions instead."
```

joostlek commented on homeassistant/components/snapcast/strings.json:
```diff
@@ -58,5 +58,15 @@
+  "issues": {
+    "deprecated_grouping_actions": {
+      "title": "Snapcast Actions Deprecated",
+      "description": "Actions 'snapcast.join' and 'snapcast.unjoin' are deprecated and will be removed in 2026.2. Use the 'media_player.join' and 'media_player.unjoin' actions instead.'"
```
```suggestion
      "description": "Actions 'snapcast.join' and 'snapcast.unjoin' are deprecated and will be removed in 2026.2. Use the 'media_player.join' and 'media_player.unjoin' actions instead."
```

mill1000 committed:
Update homeassistant/components/snapcast/config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

mill1000 committed:
Update homeassistant/components/snapcast/strings.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

mill1000 committed:
Update homeassistant/components/snapcast/strings.json

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

mill1000 committed:
Remove unnecssary config entry reloads

joostlek commented:
Can you elaborate a bit on what the code does? I will review it this week. Mostly wondering how you now do the dynamic part, as you could group 2 devices and it could create a media player entity, that would be weird.

mill1000 commented:
This code primarily implements the `media_player.async_join_players` and `media_player.async_unjoin_player` actions for Snapcast clients.

The other changes are to deprecate the existing Snapcast specific actions, and deprecate the separate media player entities for Snapcast groups.

> Mostly wondering how you now do the dynamic part, as you could group 2 devices and it could create a media player entity, that would be weird.

In it's current form the integration creates media player entities for all Snapcast clients and groups. This PR does not make any changes to that behavior.

It is useful to understand some of Snapcast's mechanics. Primarily:
- A client is always part of a group, even if it's the sole member
- Groups are dynamically created/destroyed
- Certain concepts (e.g. source) are a property of the group, not of the client.

As an example.
If we have 2 clients (Client A and Client B) that are not grouped together. HA will create 4 media player entities, 1 for each client and 1 for each group.

- Client A (`media_player.client_a_snapcast_client`)
- Group A (`media_player.client_a_snapcast_group`)
- Client B (`media_player.client_b_snapcast_client`) 
- Group B (`media_player.client_b_snapcast_group`)

If Client B is added to Group A, then Group B will be destroyed, which triggers HA to remove the media player entity for Group B.
